### PR TITLE
cache all appcontext switches

### PR DESCRIPTION
**Pull request by Wraith2**
fixes https://github.com/dotnet/SqlClient/issues/2225

Change any uncached appcontext switches to use caching like the others.

/cc @JulianRooze

### CODE REVIEW

**Review by David-Engel — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs [hunk 1]
```diff
@@ -15,8 +15,8 @@ internal static partial class LocalAppContextSwitches
 
         private static bool? s_legacyRowVersionNullBehavior;
         private static bool? s_suppressInsecureTLSWarning;
-        private static bool s_makeReadAsyncBlocking;
-        private static bool s_useMinimumLoginTimeout;
+        private static bool? s_makeReadAsyncBlocking;
```
Comment (jkotas) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs:18]: bool? is a structure with two fields. It is not updated atomically. The lock-free reads and writes below have race conditions.

If you want the reads to be lock-free, you can either:
- Change the type of these caches to `bool` and initialize them eagerly in a static constructors
or
- Change the type of these caches to atomically updated integer with 3 possible values uninitialized/false/true. Here is an example of similar pattern in dotnet/runtime: https://github.com/dotnet/runtime/blob/b88c74a109903de213b049177cbc0481a689477f/src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs#L22C22-L27

**Review by arellegue — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs (hunk 1 above)
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs]: Thanks. I've updated all the variables to use the tristate enum pattern.

**Review by David-Engel — APPROVED** [review]

**Comment by jkotas** [comment]
Failing tests:
```
  [xUnit.net 00:01:04.53]       System.InvalidCastException : Unable to cast object of type 'Tristate' to type 'System.Nullable`1[System.Boolean]'.
  [xUnit.net 00:01:04.53]       Stack Trace:
  [xUnit.net 00:01:04.53]         D:\a\_work\1\s\src\Microsoft.Data.SqlClient\tests\ManualTests\SQL\DataReaderTest\DataReaderTest.cs(330,0): at Microsoft.Data.SqlClient.ManualTesting.Tests.DataReaderTest.SetLegacyRowVersionNullBehavior(Nullable`1 value)
  [xUnit.net 00:01:04.53]         D:\a\_work\1\s\src\Microsoft.Data.SqlClient\tests\ManualTests\SQL\DataReaderTest\DataReaderTest.cs(301,0): at Microsoft.Data.SqlClient.ManualTesting.Tests.DataReaderTest.CheckLegacyNullRowVersionIsEmptyArray()
  ```

**Comment by Wraith2** [comment]
Thanks. Fixed and added note to tests and source to indicate that private reflection is being used. I wrote that test so I should have remembered.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs [hunk 2]
```diff
@@ -8,15 +8,23 @@ namespace Microsoft.Data.SqlClient
...
         internal const string MakeReadAsyncBlockingString = @"Switch.Microsoft.Data.SqlClient.MakeReadAsyncBlocking";
         internal const string LegacyRowVersionNullString = @"Switch.Microsoft.Data.SqlClient.LegacyRowVersionNullBehavior";
         internal const string SuppressInsecureTLSWarningString = @"Switch.Microsoft.Data.SqlClient.SuppressInsecureTLSWarning";
         internal const string UseMinimumLoginTimeoutString = @"Switch.Microsoft.Data.SqlClient.UseOneSecFloorInTimeoutCalculationDuringLogin";
 
-        private static bool? s_legacyRowVersionNullBehavior;
-        private static bool? s_suppressInsecureTLSWarning;
-        private static bool s_makeReadAsyncBlocking;
-        private static bool s_useMinimumLoginTimeout;
+        // this field is accesses through reflection in tests and should be be renamed or have the type changed without refactoring NullRow related tests
```
Comment (jkotas) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs:23]:
```suggestion
        // this field is accesses through reflection in tests and should be renamed or have the type changed without refactoring NullRow related tests
```

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/LocalAppContextSwitches.cs]: fixed.

**Review by jkotas — APPROVED** [review]

**Pull request merged**